nav2-usability/scenario-code/dynamic-linking/router.dart

```diff
@@ -0,0 +1,233 @@
+    // Handle '/'
+    if (uri.pathSegments.length == 0) {
+      return AppRoutePath();
+    }
+
+    // Handle '/wishlist/:id'
+    if (uri.pathSegments.length == 2) {
+      if (uri.pathSegments[0] != 'wishlist') return AppRoutePath();
+      final id = uri.pathSegments[1];
+      return AppRoutePath(id: id);
+    }
+
+    // Handle unknown routes
+    return AppRoutePath();
+  }
+
+  @override
+  RouteInformation restoreRouteInformation(AppRoutePath path) {
+    late final String location;
+    if (path.id == null) {
+      location = '/';
+    } else {
+      location = '/wishlist/${path.id}';
+    }
+    return RouteInformation(location: location);
+  }
+}
+
+class BookRouterDelegate extends RouterDelegate<AppRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<AppRoutePath> {
+  final GlobalKey<NavigatorState> navigatorKey;
+
+  AppState appState = AppState();
+
+  BookRouterDelegate() : navigatorKey = GlobalKey<NavigatorState>();
+
+  AppRoutePath get currentConfiguration {
+    final selected = appState.selectedWishlist;
+    if (selected == null) {
+      return AppRoutePath();
+    } else {
+      return AppRoutePath(id: selected.id);
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final selectedWishlist = appState.selectedWishlist;
+    return Navigator(
+      key: navigatorKey,
+      pages: [
+        MaterialPage(
+          key: ValueKey('WishlistListPage'),
+          child: WishlistListScreen(
+            wishlists: appState.wishlists,
+            onTapped: _handleTapped,
+            onCreate: (newId) {
+              setNewRoutePath(AppRoutePath(id: newId));
+            },
+          ),
+        ),
+        if (selectedWishlist != null)
+          MaterialPage(
+            key: ValueKey('WishPage'),
+            child: WishlistScreen(wishlist: selectedWishlist),
+          ),
+      ],
+      onPopPage: (route, result) {
+        if (!route.didPop(result)) {
+          return false;
+        }
+
+        // Update the list of pages by setting _selectedBook to null
+        appState.selectedWishlist = null;
+        notifyListeners();
+
+        return true;
+      },
+    );
+  }
+
+  @override
+  Future<void> setNewRoutePath(AppRoutePath path) async {
+    var pathId = path.id;
+    if (pathId == null) {
+      appState.selectedWishlist = null;
+      return;
+    }
+
+    // Create a wishlist with the given ID if none exists
+    // var hasWishlist = appState.wishlists.firstWhere((e) => e.id == path.id, orElse: () => null);
+    Wishlist? wishlist;
+    for (var w in appState.wishlists) {
+      if (w.id == path.id) {
+        wishlist = w;
+      }
+    }
+    if (wishlist == null) {
+      wishlist = Wishlist(pathId);
+      appState.wishlists.add(wishlist);
+    }
+
+    appState.selectedWishlist = wishlist;
+    notifyListeners();
+  }
+
+  void _handleTapped(Wishlist wishlist) {
+    appState.selectedWishlist = wishlist;
```

Comment:
Can we make appState.selectedWishlist setter to notify listener directly when it changes?

Reply:
Yes, I'll make AppState a ChangeNotifier